MOBT-2-Swift-Foundation.playground/Contents.swift

```diff
@@ -0,0 +1,332 @@
+    }
+}
+
+let currencyExchangeService = CurrencyExchangeConverterService()
+let currencyExchangeLocalData = CurrencyExchangeLocalData(currencyExchangeService: currencyExchangeService)
+currencyExchangeLocalData.setDataToUserDefaults()
+currencyExchangeLocalData.retrieveDataFromUserDefaults()
+
+///When we have an object from the backend, we have a possibility to create MVVM-C architecture and use data to display on UI part
+///This is example of MVVM-C structure of our project
+
+///Base coordinator conforms to different coordinators to use methods of our currency exchanges
+class BaseCoordinator {
+    private(set) weak var rootController: UINavigationController?
+    
+    init(rootController: UINavigationController) {
+        self.rootController = rootController
+    }
+    ///This is method which can be used in different coordinators to pushViewControllers, this method is not the only one which we can provide.
+    ///We can extend our base coordinator to handle different situatuations
+    func pushViewController(_ viewController: UIViewController) {
+        pushViewController(viewController)
+    }
+    ///This function needs to display that we can handle our dependencies in this class, but if we need some kind of parametres - we can create a new start function to avoid overriding this one.
+    func start() { assertionFailure("Must be overriden in subclass") }
+}
+///Example
+class CurrencyExchangeCoordinator: BaseCoordinator {
+    override func start() {
+        assertionFailure("Unreachable code")
+    }
+    
+    func start(currencyServiceLayer: CurrencyExchangeConverterService) {
+        let model = CurrencyExchangeModel(serviceLayer: currencyServiceLayer)
+        let viewModel = CurrencyExchangeViewModel(model: model)
+        let viewController = CurrencyExchangeViewController(viewModel: viewModel)
+        pushViewController(viewController)
+    }
+}
+
+///This peace of code we use to override basic functions of ViewController, just conform it in your viewController
+public protocol ControlSetup {
+    func setupSubviews()
+    func setupAutoLayout()
+    func setupStyle()
+    func setupActions()
+}
+
+public extension ControlSetup {
+    func setupActions() { }
+    
+    func controlSetup() {
+        setupSubviews()
+        setupAutoLayout()
+        setupStyle()
+        setupActions()
+    }
+}
+
+///ViewController can help us to introduce to user our CurrencyExchange using data from viewModel
+///In this case we do not need any base classes because we don't need to handle some errors or etc.
+///We can create 'Control Setup' method to easilly add base functions of our viewController
+
+protocol CurrencyExchangeStyling {
+    ///declare variables to style your viewController
+}
+
+struct CurrencyExchangeStyle: CurrencyExchangeStyling {
+    ///setup your style here
+}
+
+//Example
+class CurrencyExchangeViewController: UIViewController, ControlSetup {
+    
+    var viewModel: CurrencyExchangeViewModel
```

Comment:
Do we expect this properties to be changed during the controller lifetime? If not — they should be let. Do we expect anyone to access these values outside of controller scope? If not — they should be private (like most of properties by default)